### PR TITLE
Replace set_last_error with set_last_error_and_return

### PR DESCRIPTION
**Pull request by noahmbright**
Took care of the simple patterns. Other patterns involved setting an error and then using `write_int` or setting metadata and returning -1. Unsure if those are in the scope of this change

Looks like this has conflicts with #3779, so I can update when how to handle that is decided. 

Part of https://github.com/rust-lang/miri/issues/3930.

### CODE REVIEW

File: src/shims/unix/fs.rs
Comment (oli-obk) [inline src/shims/unix/fs.rs:1]: Where did the stray formatting changes come from?

File: src/shims/unix/fd.rs
Comment (oli-obk) [inline src/shims/unix/fd.rs:557]: Use `LibcError("EBADF")`

Comment (oli-obk) [reply src/shims/unix/fd.rs:557]: Also in other places

**Review by oli-obk — CHANGES_REQUESTED** [review]
Please revert the formatting changes unless they are from `./miri fmt` (you may need to configure your vscode correctly with the settings from `CONTRIBUTING.md`)

Also use `LibcError("foobar")` everywhere while you're doing this change.

**Comment by oli-obk** [comment]
Ignore my PR, i'll rebase at some point

File: src/shims/unix/fs.rs
Comment (RalfJung) [inline src/shims/unix/fs.rs:604]: This returns an `i64`, not an `i32`, so your patch introduces an ICE here! So please add a new `set_last_error_and_return_i64` instead and use that here, or leave this to a future PR.

Please make sure to always maintain the original type, it's quite important to not screw this up and because this is "dynamic" (it's types of the interpreted program, not of the interpreter), even `./miri check` will not catch this.

File: src/shims/unix/fs.rs
Comment (RalfJung) [inline src/shims/unix/fs.rs:594]: More incorrect `i64` -> `i32` changes

File: src/shims/unix/fs.rs
Comment (RalfJung) [inline src/shims/unix/fs.rs:1030]: This returned a null pointer before and now you made it return -1. That's definitely wrong, please keep the old code here.

File: src/shims/unix/fs.rs
Comment (tiif) [inline src/shims/unix/fs.rs:1422]: It seems like previously it returns ``Ok(-1)`` but now it returns ``Ok(Ok(-1))``?

**Comment by RalfJung** [comment]
Not sure why the comments above have all been marked as resolved -- in the PR they are definitely still unresolved.

@rustbot author

**Comment by noahmbright** [comment]
> Not sure why the comments above have all been marked as resolved -- in the PR they are definitely still unresolved.

Yes, sorry for the wait. How do you want to handle the `InterpResult<'tcx, i64>` vs . `InterpResult<'tcx, Scalar>` issue in my comment above? Once I fix that I'll push the update.

**Comment by RalfJung** [comment]
Which comment?

File: src/shims/unix/fs.rs
Comment (RalfJung) [reply src/shims/unix/fs.rs:1422]: For now just change this to `this.set_last_error(LibcError("EACCES"))?;` but leave the `return` alone.

**Comment by noahmbright** [comment]
Again sorry about the wait and the careless mistakes, but I added two commits, one a small formatting fix, and in the other I added `set_last_error_and_return_i64` and fixed the type issues I introduced. Diff is smaller now too so hopefully this will be easier to re-review.

**Comment by RalfJung** [comment]
There are unfortunately conflicts with the master branch, could you take care of them?
Then write @`rustbot ready`  to initiate the next round of review.

**Comment by noahmbright** [comment]
> There are unfortunately conflicts with the master branch, could you take care of them with a rebase? Then write @`rustbot ready` to initiate the next round of review.

Of course. All the conflicts are something like this:
```
<<<<<<< unix_shims
            return this.set_last_error_and_return_i32(ErrorKind::PermissionDenied);
=======
            this.set_last_error(ErrorKind::PermissionDenied)?;
            return interp_ok(Scalar::from_i32(-1));
>>>>>>> master
```

I can wrap `interp_ok(this.set_last_error_and_return_*(e))` where these all occur, and we'll have an `interp_ok(InterpResult<'tcx, Scalar>)`. Does that sound alright for the resolution?

**Comment by RalfJung** [comment]
`return this.set_last_error_and_return_i32(ErrorKind::PermissionDenied);` looks like the right resolution for that example.

**Comment by noahmbright** [comment]
> `return this.set_last_error_and_return_i32(ErrorKind::PermissionDenied);` looks like the right resolution for that example.

I agree, after seeing the new changes while rebasing. @rustbot ready

File: src/shims/unix/linux/epoll.rs
Comment (RalfJung) [inline src/shims/unix/linux/epoll.rs:372]: This could use `LibcError`

**Review by RalfJung — COMMENTED** [review]
Thanks, this all looks correct but there is one spot where we can make it a bit more compact. :)

Please also squash the commits (with `git rebase --keep-base`).

**Comment by noahmbright** [comment]
> Thanks, this all looks correct but there is one spot where we can make it a bit more compact. :)

Good catch. Thank you for the close reviews, this should have been much simpler. If I'm still welcome, I'll make sure my next PR goes more smoothly

**Comment by RalfJung** [comment]
Looks good, thanks for the PR!
@bors r+

> Thank you for the close reviews, this should have been much simpler. If I'm still welcome, I'll make sure my next PR goes more smoothly

Sure thing, hickups happen. :) Just be careful and ask us rather than making guesses if you're not sure about something. :D When submitting a PR, it is a very good idea to list which things in the PR you are least sure about -- that greatly helps focus the review effort.

**Comment by RalfJung** [comment]
That's odd... @rust-lang/infra bors has a problem if we leave a "changes required" review, and then the changes are done? CI is green but the PR has not been pushed to master.

**Review by RalfJung — APPROVED** [review]
Let's see if this helps github.

**Comment by Kobzol** [comment]
Wtf, I never saw this error before.. Anyway, what do you think about switching miri from bors to merge queues? xD

**Comment by RalfJung** [comment]
> Wtf, I never saw this error before.. Anyway, what do you think about switching miri from bors to merge queues? xD

Yeah someone just needs to put in the work for that and I don't currently have the time.^^

**Comment by RalfJung** [comment]
@oli-obk I think we need to you to send an "approving" review or so, otherwise this PR can't land...

Is that really how github reviews work? Nobody else can approve a PR after a team member left a "changes required" review? That's a complete disaster.

**Comment by oli-obk** [comment]
you should be able to dismiss my review in the github ui (i approved it now, but for the future)

@bors r+

**Comment by RalfJung** [comment]
Where is the "dismiss review" button?

**Comment by RalfJung** [comment]
Ah I think I found it. That's quite well-hidden...

**Pull request merged**
